Source/monster.cpp

```diff
@@ -2385,199 +2215,149 @@ void __fastcall M_DiabloDeath(int i, BOOL sendmsg)
 
 void __fastcall M2MStartHit(int mid, int i, int dam)
 {
-	int v3; // edi
-	int v4; // ebx
-	int v5; // esi
-	CMonster *v6; // eax
-	char v7; // al
-	CMonster *v8; // eax
-	int v9; // ecx
-	int v10; // eax
-	int v11; // [esp+Ch] [ebp-4h]
-
-	v3 = mid;
-	v4 = i;
-	v11 = i;
-	if ( (unsigned int)mid >= MAXMONSTERS )
+	if ( (DWORD)mid >= MAXMONSTERS ) {
 		TermMsg("Invalid monster %d getting hit by monster", mid);
-	v5 = v3;
-	if ( !monster[v3].MType )
-		TermMsg("Monster %d \"%s\" getting hit by monster: MType NULL", v3, monster[v5].mName);
-	if ( v4 >= 0 )
-		monster[v4].mWhoHit |= 1 << v4;
-	delta_monster_hp(v3, monster[v5]._mhitpoints, currlevel);
-	NetSendCmdParam2(FALSE, CMD_MONSTDAMAGE, v3, dam);
-	PlayEffect(v3, 1);
-	v6 = monster[v5].MType;
-	if ( v6->mtype >= MT_SNEAK && v6->mtype <= MT_ILLWEAV || dam >> 6 >= SLOBYTE(monster[v5].mLevel) + 3 )
-	{
-		if ( v11 >= 0 )
-			monster[v5]._mdir = ((unsigned char)monster[v11]._mdir - 4) & 7;
-		v7 = v6->mtype;
-		if ( v7 == 39 )
-		{
-			M_Teleport(v3);
-		}
-		else if ( v7 >= MT_NSCAV && v7 <= MT_YSCAV )
-		{
-			_LOBYTE(monster[v5]._mgoal) = 1;
+	}
+
+	if ( !monster[mid].MType ) {

```

Comment:
Explicit checks for NULL, zero and false do not match the (agreed upon) WebKit style guide: https://webkit.org/code-style-guidelines/#null-false-and-zero

Additionally I don't see why they should be used unless there is a reason beyond personal preference, they just create more work.

Reply:
They increase readability / unambiguousness, I mean here anyone always know MType is a pointer (not an int eg.) without jumping to the declaration.

Reply:
That is still a case of preference, we are meant to be restoring the code as it originally was so there should be evidence of it being one way or the other before forcing it to be adopted.

Reply:
Ah, right.  Is there any evidence of explicit NULL checks?

Reply:
@squidcc the webkit style was just adopted just as basis for clang-format. Since devilution never had a proper style guide AJenbo, mewmew and me wrote this as starting point: https://github.com/diasurgical/devilution/wiki/Code-style-guide
If you have any additions or notes about them please share them! :)

Reply:
I could argue against pretty much everything in those guidelines; the DWORD casts are merely an optimisation to avoid having to check for less-than-zero, evidence of pre- and post-fix operators is in the assembly, etc.

I think I'm wasting my time though.

Reply:
>I could argue against pretty much everything in those guidelines

Yeah, they are definitely opinionated (as are yours).

> the DWORD casts are merely an optimisation to avoid having to check for less-than-zero

Yes, and that's why it hints towards the usage of `DWORD` instead of `unsigned int`.

> evidence of pre- and post-fix operators is in the assembly, etc.

as stated in the code style guide, as long as the assembly is unaffected, we prefer postfix. If you encounter places where a prefix increment is needed to generate the right instruction ordering, go for it! (I think I had one of such cases somewhere)

> There are also very relevant points completely absent (variables declared in the middle of functions should not be allowed but are rampantly used).

Since all code is and was compiled as C++ code, not as raw C code, I don't see the problem with it, as long as binary accuracy is not affected by that and other info (like line numbers) doesn't contradict that either.

> These "guidelines" directly conflict with what galaxyhaxz told me

How do the guidelines in the wiki contradict with the info in the linked comment by galaxy? He exactly outlined the types used in the wiki. The only thing missing from that is the info about the hungarian notation prefixes, which I'll add now. Thanks for pointing to that!

Reply:
I think we are all just trying our best.

As I understood the C vs C++ compiler issue; the header only indicates that some part was compiled as C, but not what part. My guess would be that this would then mostly apply to render.cpp, especially as it was compiled out of order from the other files. I think we have also seen indication that some files defiantly used C++ features? Again if my understanding of the issue changes I'm more then happy to go back and change the code.

A lot of the deliberation over code style etc. happens on the chat so it would be really good if you could join in there, again I don't think it's really fair to have a big discussion about things on line 2222 of my PR :)

Reply:
> When you see decompiled code like this:

That code would definitely be an exception (is that actual code somewhere? If so, please point me to it and I'll change that). Especially in (but not exclusive to) code for which the declared variables are known introducing a new variable just to adhere the guide doesn't make sense.

> arbitrarily forbid

The style guide is just that, a guide. If there's a good reason to do something else, go for it, but questions about these parts may come of course.

> uncommon practices like explicit comparisons with zero

That's 100% subjective. Some people prefer `ptr == NULL`, some `NULL == ptr`, some `!ptr`, etc. Each of these can be reasoned for and against. You asked for a code style guide quite a long time ago, so we got together and addressed the main friction points for now.

The WebKit style guide was just used as base for the clang-format rules, being relatively close to IDA's output, nothing more, nothing less.

> if the code matches, doesn't have anything hideously ugly (i.e. casts of different sized pointers, gotos, local function declarations...) and doesn't conflict with known information, it should be accepted.

The guide was set up to be able to have some consistency. The code style can vary greatly between people, so setting a few guidelines makes sense here.

> That is precisely the problem: nearly every file was compiled as C code. There is rock-solid evidence of this in the Rich header, which has been pointed out and ignored.

That's a different thing then. I've not been able to track the progress in devilution quite as much in the last few weeks. Could you point me to the discussion about this? If that's the case, I think we should alter the VC project files and Makefiles to reflect that information, moving the declarations up while we're at it. Would you then mind creating an issue for that?

> His instruction (which I agree with because it actually makes sense) was to only use the winapi types when there is evidence of them being used, otherwise use natural types. That's not what the style guide says; it declares they must be used for all unsigned variables. unsigned char was even given as an acceptable example, so I don't know how you don't see that as a contradiction.

For the sake of consistency it makes sense to use the same set of general types throughout the project. I understood the `unsigned char` in galaxy's answer as just being for completely unknown data, for which it makes sense to just leave it as generated.

Again it's a style *guide*, not a style *decree*. It's just meant to add a bit of consistency to the code and prevent too many of these discussions.

Reply:
>  I don't think it's really fair to have a big discussion about things on line 2222 of my PR :)

Funny you should mention this specific line number, I think it's the perfect place to have a discussion about the merits of enforcing unnecessary additional code and the inherent possibility for associated mistakes...